hermes-frontend/src/main/java/pl/allegro/tech/hermes/frontend/producer/kafka/MultiDatacenterMessageProducer.java

```diff
@@ -81,6 +109,90 @@ private void send(KafkaMessageSender<byte[], byte[]> sender,
+        } else {
+            send(sender, producerRecord, callback, cachedTopic, message);
+        }
+    }
+
+    private void scheduleChaosExperiment(ChaosExperiment experiment,
+                                         KafkaMessageSender<byte[], byte[]> sender,
+                                         ProducerRecord<byte[], byte[]> producerRecord,
+                                         SendCallback callback,
+                                         CachedTopic cachedTopic,
+                                         Message message) {
+        try {
+            chaosScheduler.schedule(() -> {
+                if (experiment.completeWithError()) {
+                    var datacenter = sender.getDatacenter();
+                    var exception = new ChaosException(datacenter, experiment.delayInMillis(), message.getId());
+                    callback.onUnpublished(message, cachedTopic, datacenter, exception);
+                } else {
+                    send(sender, producerRecord, callback, cachedTopic, message);
+                }
+            }, experiment.delayInMillis(), TimeUnit.MILLISECONDS);
+        } catch (RejectedExecutionException e) {
+            logger.warn("Failed while scheduling chaos experiment. Sending message to Kafka.", e);
+            send(sender, producerRecord, callback, cachedTopic, message);
+        }
+    }
+
+    private Map<String, ChaosExperiment> createChaosExperimentsPerDatacenter(Topic topic) {
+        PublishingChaosPolicy chaos = topic.getChaos();
+        return switch (chaos.mode()) {
+            case DISABLED -> Map.of();
+            case GLOBAL -> {
+                Map<String, ChaosExperiment> experiments = new HashMap<>();
+                ChaosPolicy policy = chaos.globalPolicy();
+                boolean enabled = computeIfShouldBeEnabled(policy);
+                for (String datacenter : kafkaMessageSenders.getDatacenters()) {
+                    experiments.put(datacenter, createChaosExperimentForDatacenter(policy, enabled));
+                }
+                yield experiments;
+            }
+            case DATACENTER -> {
+                Map<String, ChaosExperiment> experiments = new HashMap<>();
+                Map<String, ChaosPolicy> policies = chaos.datacenterPolicies();
+                for (String datacenter : kafkaMessageSenders.getDatacenters()) {
+                    ChaosPolicy policy = policies.get(datacenter);
+                    boolean enabled = computeIfShouldBeEnabled(policy);
+                    experiments.put(datacenter, createChaosExperimentForDatacenter(policy, enabled));
+                }
+                yield experiments;
+            }
+        };
+    }
+
+    private boolean computeIfShouldBeEnabled(ChaosPolicy policy) {
+        if (policy == null) {
+            return false;
+        }
+        return random.nextInt(100) < policy.probability();
```

Comment:
Shouldn't we use `ThreadLocalRandom` here as well? this will be executed from many threads concurrently, and guided by the java doc of `java.util.Random` this is not adviced:

> nstances of java. util. Random are threadsafe. However, the concurrent use of the same java. util. Random instance across threads may encounter contention and consequent poor performance. Consider instead using java. util. concurrent. ThreadLocalRandom in multithreaded designs.

Reply:
Correct, fixed.